plugins/performance-lab/includes/admin/plugins.php

```diff
@@ -71,16 +71,29 @@ function perflab_query_plugin_info( string $plugin_slug ) {
 		return new WP_Error( 'no_plugins', __( 'No plugins found in the API response.', 'performance-lab' ) );
 	}
 
-	$plugins            = array();
-	$standalone_plugins = array_merge(
-		array_flip( perflab_get_standalone_plugins() ),
-		array( 'optimization-detective' => array() ) // TODO: Programmatically discover the plugin dependencies and add them here. See <https://github.com/WordPress/performance/issues/1616>.
-	);
-	foreach ( $response->plugins as $plugin_data ) {
-		if ( ! isset( $standalone_plugins[ $plugin_data['slug'] ] ) ) {
+	$plugins      = array();
+	$plugin_queue = perflab_get_standalone_plugins();
+
+	// Index the plugins from the API response by their slug for efficient lookup.
+	$plugins_by_slug = array_column( $response->plugins, null, 'slug' );
+
+	// Start processing the plugins using a queue-based approach.
+	while ( ! empty( $plugin_queue ) ) {
```

Comment:
Why use a `while` loop here? I think it would be more intuitive to use a `foreach` loop like before, that iterates through the `perflab_get_standalone_plugins()` items. What we need to iterate over here is known, so I don't see a need for using `while`.

Then in the loop you could populate `$plugins` with the data, and if there are dependencies (`requires_plugins`), add the data for them to the array as well (unless it was already added).

Reply:
The reason was if `foreach` is used then there is need of the nested `foreach` for handling the dependencies which will have the same logic of outer `foreach` to add it into the `$plugins`. So by using the `while` we can eliminate the duplicate logic. 

Also I tried using `foreach` with appending elements to the array but in PHP `foreach` works by creating an internal copy of the array's state at the beginning of the loop. Which means if we add other elements to that array it will not iterate over them. 
This could also can be solved by using the simple `for` loop but `while` looked more clean.

Reply:
Ah I see, because there could in theory be dependencies that have dependencies themselves. While we don't currently have that, that's a good point, so let's keep it as is.